### ISSUE TITLE
bug in _evaluate_patterns 

### ISSUE THREAD

**Issue by idsvandermolen** [ansible/ansible#26480]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
inventory manager

##### ANSIBLE VERSION

```
ansible 2.3.1.0
  config file = /Users/idsvandermolen/projects/ansible-devops/ansible.cfg
  configured module search path = [u'./library']
  python version = 2.7.12 (default, Sep 28 2016, 18:41:32) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.38)]
```

##### OS / ENVIRONMENT

macOS Sierra 10.12.5

##### SUMMARY

There's a bug in _evaluate_patterns method. It checks for existence of pattern in `self._hosts_cache` 2.3.1.0) or `self._inventory.hosts` (https://github.com/ansible/ansible/blob/392934f1b9cf85672be927a63bfb8702dca3f8e5/lib/ansible/inventory/manager.py#L381) but fails to check if the entry is None

##### STEPS TO REPRODUCE

We use a playbook with multiple host inventory files. A host pattern in the playbook might be available in 1 host inventory file, but not the other.  Running the playbook with the inventory file that is missing the host will result in ansible throwing an exception.

##### EXPECTED RESULTS

It should just skip the task

##### ACTUAL RESULTS

```
skipping: no hosts matched
ERROR! Unexpected Exception: 'NoneType' object has no attribute 'name'
the full traceback was:

Traceback (most recent call last):
  File "/usr/local/bin/ansible-playbook", line 109, in <module>
    exit_code = cli.run()
  File "/usr/local/lib/python2.7/site-packages/ansible/cli/playbook.py", line 154, in run
    results = pbex.run()
  File "/usr/local/lib/python2.7/site-packages/ansible/executor/playbook_executor.py", line 125, in run
    all_vars = self._variable_manager.get_vars(loader=self._loader, play=play)
  File "/usr/local/lib/python2.7/site-packages/ansible/vars/__init__.py", line 236, in get_vars
    include_delegate_to=include_delegate_to,
  File "/usr/local/lib/python2.7/site-packages/ansible/vars/__init__.py", line 429, in _get_magic_variables
    variables['ansible_play_hosts_all'] = [x.name for x in self._inventory.get_hosts(pattern=pattern, ignore_restrictions=True)]
  File "/usr/local/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 248, in get_hosts
    hosts = self._evaluate_patterns(patterns)
  File "/usr/local/lib/python2.7/site-packages/ansible/inventory/__init__.py", line 352, in _evaluate_patterns
    to_append = [ h for h in that if h.name not in [ y.name for y in hosts ] ]
AttributeError: 'NoneType' object has no attribute 'name'
```

The fix in 2.3.1.0 is to patch `ansible/inventory/__init__.py`:
```
     for p in patterns:
            # avoid resolving a pattern that is a plain host
            if p in self._hosts_cache:
                h = self.get_host(p)
                if h is not None:
                    hosts.append(h)
                #hosts.append(self.get_host(p))
            else:
```
Similar fix could be applied to latest version https://github.com/ansible/ansible/blob/392934f1b9cf85672be927a63bfb8702dca3f8e5/lib/ansible/inventory/manager.py#L381
```
for p in patterns:
    # avoid resolving a pattern that is a plain host
    if p in self._inventory.hosts:
        h = self._inventory.get_host(p)
        if h is not None:
            hosts.append(h)
        #hosts.append(self._inventory.get_host(p))
    else:
```

**Comment by nitzmahone** [comment]
Can you include a small sample playbook and inventory that exhibits the behavior? As I'm understanding it, I'm not able to reproduce the problem on 2.3.1. We'll need to see exactly what you're trying to do.

**Comment by idsvandermolen** [comment]
I haven't been able to narrow it down to a reproducible playbook yet, it appears in our company playbooks. Basically we have something like this:
hosts:
```
[group1]
host1
```
base.yml:
```
---
- hosts: missinghost
  tasks:
    - name: Base missinghost
      debug: msg="Hello from base for missinghost"

- hosts: group1
  tasks:
    - name: Base group1
      debug: msg="Hello from base for group1"
```
playbook.yml:
```
---
- include: base.yml
- hosts: missinghost, group1
  tasks:
    - name: Test
      debug: msg="Hello from test"
```
NOTE: missinghost is NOT in the hosts file, and above playbooks do not reproduce the issue.
We would get the exception when we ran the playbook like this:
`ansible-playbook playbook.yml -i hosts -l host1`
At some point the `missinghost` would be evaluated (maybe through `get_host()`) and add a entry with value `None` in `_hosts_cache`. This breaks `_evaluate_patterns` at a later point because that method does not check for `None` values in `_hosts_cache`, throwing the Exception. 
Our current work-around is to rewrite the playbook like this:
```
---
- include: base.yml
- hosts: missinghost
  tasks:
    - name: Test
      debug: msg="Hello from test"
---
- include: base.yml
- hosts: group1
  tasks:
    - name: Test
      debug: msg="Hello from test"
```
This way `missinghost` is skipped and doesn't end up in the cache with a None value.
I'll try to come up with a minimal playbook that reproduces the issue, but currently I do not know how hosts end up in the `_hosts_cache`.

**Comment by idsvandermolen** [comment]
Narrowed it down a bit, this will trigger exception when run like this:
`ansible-playbook playbook.yml -i hosts -l localhost -v`
hosts:
```
[group1]
localhost
```
playbook.yml:
```
---
- include: base.yml
- hosts: missinghost, group1
  tasks:
    - name: Test
      debug: msg="Hello from test"
```
base.yml:
```
---
- hosts: missinghost
  tasks:
    - name: Base missinghost
      debug: msg="Hello from base for missinghost"

- hosts: group1
  tasks:
    - name: Delegate task
      debug: msg="Delegate this to missinghost"
      delegate_to: missinghost

    - name: Base group1
      debug: msg="Hello from base for group1"
```
It doesn't trigger exception when you modify playbook.yml to:
```
---
- include: base.yml
- hosts: missinghost
  tasks:
    - name: Test
      debug: msg="Hello from test"

- hosts: group1
  tasks:
    - name: Test
      debug: msg="Hello from test"
```

**Comment by sivel** [comment]
This issue was resolved in 8f97aef1a365cbbbb822d6d09f96af17a076b295 and subsequently included in the 2.4 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
